### PR TITLE
Rewrite with `hikari`

### PR DESCRIPTION
**Pull request by AlexanderHOtt**
This is a basic structure for the discord bot. None of the features have been implemented yet.

### CODE REVIEW

File: discord-bot/bot/config.py
Comment (andreaskoepf) [inline discord-bot/bot/config.py:16]: Could you please use a [pydantic settings](https://docs.pydantic.dev/usage/settings/) class here?

File: discord-bot/flake8-requirements.txt
Comment (andreaskoepf) [inline discord-bot/flake8-requirements.txt:6]: Is this necessary? we have project-wide settings in place via pre-commit...

File: discord-bot/noxfile.py
Comment (andreaskoepf) [inline discord-bot/noxfile.py:6]: we use pre-commit .. please do not add other project-specific style enforcement tools!

File: discord-bot/pyproject.toml
Comment (andreaskoepf) [inline discord-bot/pyproject.toml:5]: same goes here .. please no new toml file!

File: discord-bot/bot/extensions/example.py
Comment (andreaskoepf) [inline discord-bot/bot/extensions/example.py:3]: why do we need this 406 lines example file in the project? Could you maybe link this in a readme?

File: discord-bot/bot/config.py
Comment (andreaskoepf) [inline discord-bot/bot/config.py:35]: see example from pydantic docs...
```
class Settings(BaseSettings):
    numbers: list[int]

    class Config:
        @classmethod
        def parse_env_var(cls, field_name: str, raw_val: str) -> Any:
            if field_name == 'numbers':
                return [int(x) for x in raw_val.split(',')]
            return cls.json_loads(raw_val)
```

**Comment by andreaskoepf** [comment]
We will keep this PR open until we hit feature parity.

File: discord-bot/bot/extensions/tasks.py
Comment (andreaskoepf) [inline discord-bot/bot/extensions/tasks.py:274]: this is not always 5, it depends on the number of elements in the task.

File: discord-bot/bot/extensions/tasks.py
Comment (andreaskoepf) [inline discord-bot/bot/extensions/tasks.py:219]: There is `scale: RatingScale` that defines min and max for the rating.

File: discord-bot/bot/extensions/tasks.py
Comment (andreaskoepf) [inline discord-bot/bot/extensions/tasks.py:81]: these hard-coded message strings should ideally be moved outside the main code files for people to edit.

**Comment by AlexanderHOtt** [comment]
I think the bot now has feature parity, correct me if I'm wrong.

# Tasks

- [x] `/work` to start a new task
(In the server)
image
(In DMs)
image
- [x] User can select which type of task (`rate_summary` and `summarize_story` are not implemented yet)
image
- [x] Correct API interaction though the api client.
- [x] Timeout/cancel handling (Accepting the task is a 1 min timeout, responding to the task is a 1 hour timeout)
image

- [x] Embed templates (these are not "quickly editable" by people who unfamiliar with code, but they look a lot better)
image


# Text Labels

- [ ] text labels through a message context menu
I'm not sure how to report these to the backend yet, but the frontend is working.
I chose to let the user edit what text they are labeling, but I do keep track if they edit the message. Should I change this?
image
image
image
image

# Log Channel Configuration
- [x] `/settings log_channel` to set the log channel the bot sends completed task messages to.
image
- [x] `/settings get` to get the current server settings
image


I'll start cleaning up the code, here are my plans so far, but please mention anything that is missing.

Immediate
- [ ] address `TODO` comments
- [ ] backend text labeling
- [ ] switch logging to `loguru` and add more + better log messages
- [ ] switch bot config to `pydantic`
- [ ] switch nox sessions to precommit-hooks
- [ ] move example.py to a markdown file
- [ ] document codebase structure
- [ ] docker setup (maybe?)

Quality of life / UI
- [ ] improve bot message wording
- [ ] switch most (if not all) bot messages to embeds
- [ ] move all user facing messages (embeds) to a central place
- [ ] add better images to embeds
- [ ] `/help`
- [ ] `/tutorial`

Waiting on
- [ ] leaderboard / points

**Comment by AlexanderHOtt** [comment]
@yk you created #40, did you mean that users can add text labels to task responses or to _any_ message?

File: discord-bot/bot/settings.py
Comment (andreaskoepf) [inline discord-bot/bot/settings.py:15]: Could you please add `case_sensitive = False`, I guess it is default in pydantic but just to be explicit (since unix env names normally capitalized).

**Review by andreaskoepf — COMMENTED** [review]
LGTM. API settings should be read from conf and I am still not sure about the large example file. Is this something special that you wrote specific for the project or is it something which is on the net and we could simply link to it? Currently it looks like a bit of junk (e.g. we also do not include full manual and examples to all other libs we use).

The comments that I made regarding the messages, external string-table/templates etc. could be addressed later after merging this. We should review the overall interaction concept first.

**Comment by AlexanderHOtt** [comment]
About the example file, it’s something I wrote to help new contributors get up to speed quickly with hikari. There are some examples online, but they are spread all over the place. I want to include it somewhere, even if that is in the project wiki or a gist.

File: discord-bot/bot/settings.py
Comment (yk) [reply discord-bot/bot/settings.py:15]: 1. I don't think there's a need to inherit form `BaseSettings.Config`, I think they're merged automatically
2. maybe it's worth (in the future) to make a base settings class of our own where we have things like case sensitivity and the presence of a `.env` file in.

**Comment by yk** [comment]
> About the example file, it’s something I wrote to help new contributors get up to speed quickly with hikari. There are some examples online, but they are spread all over the place. I want to include it somewhere, even if that is in the project wiki or a gist.

I think the ideal way would be to put it into a gist, or maybe you already have it somewhere in an open repo, then link to that in `discord-bot/README.md`. Although I'm not particularly against having examples, but ideally these come in the form of unit tests, which I can see is a bit tricky with discord. So no strong opinions, but if the file exists somewhere already or if a gist is fine, then I think a link would be perfect. Otherwise including it seems ok.

File: discord-bot/bot/extensions/tasks.py
Comment (AlexanderHOtt) [reply discord-bot/bot/extensions/tasks.py:216]: the `tasks.py` file was for testing different types of user input. If we decide to go with select menus for rating, I will update the code in `work.py`. I'll rename `task.py` to `task_tests.py` to make that more clear.

File: discord-bot/bot/settings.py
Comment (AlexanderHOtt) [reply discord-bot/bot/settings.py:17]: > 1. I don't think there's a need to inherit form `BaseSettings.Config`, I think they're merged automatically
> 2. maybe it's worth (in the future) to make a base settings class of our own where we have things like case sensitivity and the presence of a `.env` file in.

When I don't inherit, I get an error from my type-checker
```
"Config" overrides symbol of same name in class "BaseSettings"
  "discord-bot.bot.settings.Settings.Config" is incompatible with "pydantic.env_settings.BaseSettings.Config"
  Type "Type[Config]" cannot be assigned to type "Type[Config]"

Pylance[reportIncompatibleVariableOverride]
```
The message is a bit cryptic, but because we are subclassing `pydantic.BaseSettings`, we need to keep the types the same (including `class Config` which inherits from `pydantic.BaseConfig`, but we can't just inherit from `pydantic.BaseConfig`, we have to inherit from `pydantic.BaseSettings.Config`). This "defining a class inside a class" always confused me a bit, but I guess it works. I could add a `# type: ignore` comment, but inheriting make more sense to me.

**Review by andreaskoepf — APPROVED** [review]
OK, ready for bot replacement.

**Pull request merged**
